assisted-events-scrape/storage/elasticsearch_storage.py

```diff
@@ -0,0 +1,76 @@
+from typing import List, Callable, Iterable
+from utils import log
+from config import ElasticsearchConfig
+from elasticsearch import Elasticsearch, helpers
+from elasticsearch.exceptions import NotFoundError
+
+DEFAULT_SCAN_SIZE = 500
+DEFAULT_SCROLL_WINDOW = '5m'
+
+
+class ElasticsearchStorage:
+    @staticmethod
+    def create() -> 'ElasticsearchStorage':
+        config = ElasticsearchConfig.create_from_env()
+        http_auth = None
+        if config.username:
+            http_auth = (config.username, config.password)
+        es_client = Elasticsearch(config.host, http_auth=http_auth)
+        return ElasticsearchStorage(es_client)
+
+    def __init__(self, es_client: Elasticsearch):
+        self._es_client = es_client
+
+    def store_changes(self, index: str, documents: List[dict], id_fn: Callable[[dict], str], filter_by: dict = None):
+        actions = self._get_new_documents_actions(
+            index=index,
+            documents=documents,
+            id_fn=id_fn,
+            filter_by=filter_by)
+        return helpers.bulk(self._es_client, actions)
+
+    def _get_new_documents_actions(self, index: str, documents: List[dict],
+                                   id_fn: Callable[[dict], str], filter_by: dict) -> Iterable[dict]:
+        all_ids = set()
+        existing_docs = self.scan(index=index, filter_by=filter_by)
+
+        for d in existing_docs:
+            all_ids.add(d["_id"])
+        for d in documents:
+            doc_id = id_fn(d)
+            if doc_id not in all_ids:
+                yield {
+                    "_index": index,
+                    "_id": doc_id,
+                    "_source": d,
+                    "_op_type": "index"
+                }
+
+    # pylint: disable=dangerous-default-value
+    def scan(self, index: str, filter_by: dict = None, source: list = [""],
+             scroll: str = DEFAULT_SCROLL_WINDOW, size: int = DEFAULT_SCAN_SIZE) -> Iterable[dict]:
+        body = {"query": {"match_all": {}}}
+        if filter_by is not None:
+            body = {"query": {"term": filter_by}}
+
+        hits = []
+
+        try:
+            docs = self._es_client.search(
+                index=index,
+                body=body,
+                scroll=scroll,
+                size=size,
+                _source=source
+            )
+            hits = docs["hits"]["hits"]
+            scroll_id = docs["_scroll_id"]
+            yield from hits
+        except NotFoundError:
+            log.warning(f"Could not scan index {index}: not found")
+
+        while len(hits) > 0:
+            docs = self._es_client.scroll(scroll_id=scroll_id, scroll=scroll)
+            scroll_id = docs["_scroll_id"]
+            hits = docs["hits"]["hits"]
+            yield from hits
```

Comment:
there is ``elasticsearch.helpers.scan`` for doing this logic, no?

Reply:
Yes, if only accepted `source` as parameter I would've used that.
But you could not know as there was documentation, this is now documented :)